lib/repo/repository.rb

```diff
@@ -204,6 +204,47 @@ def self.delete_bulk_permissions(groups, user_ids)
       raise NotImplementedError, "Repository.delete_bulk_permissions: Not yet implemented"
     end
 
+    # Builds a hash of all repositories and users allowed to rw to them
+    def self.get_all_permissions
+
+      permissions = {}
+      admins = Admin.all
+      admins = admins.map(&:user_name)
+      tas = Ta.all
+      tas = tas.map(&:user_name)
+      non_student_repos = Group.all
+      non_student_repos = non_student_repos.map(&:repository_name)
+      # Permission subtleties:
+      # 1) a repository is associated with a Group, but..
+      # 2) ..students are associated with a Grouping (an "instance" of Group for a specific Assignment)
+      # That creates a problem since authentication in svn/git is at the repository level, while Markus handles it at
+      # the assignment level, allowing the same Group repo to have different students according to the assignment.
+      # The two extremes to implement it are union (permissive) or intersection (restrictive), but we are going to take
+      # a last-deadline approach instead, where we assume that the valid students at any point in time are the ones
+      # valid for the last assignment due.
+      # (Basically, it's nice for a group to share a repo among assignments, but at a certain point during the course
+      # we may want to add or [more frequently] remove some students from it)
+      assignments = Assignment.order(due_date: :desc)
```

Comment:
Ok I get it.
If I go as deep as possible, does it interact well with the various subqueries, e.g. assignment.valid_groupings and valid_grouping.accepted_students?